client/lib/navigate/index.ts

```diff
@@ -13,9 +13,17 @@ function isPageRegistered() {
 	return !! String( page.base() );
 }
 
+// Check if the path is within Calypso's scope.
+// For example, the path "/home" is within Calypso's scope.
+// The path "/setup" is not within Calypso's scope. Its part of the Stepper framework.
+function isPathOutOfScope( redirectPath: string ): boolean {
+	const paths = [ '/setup' ];
+	return paths.some( ( path ) => redirectPath.startsWith( path ) );
+}
+
 export function navigate( path: string ): void {
 	if ( isSameOrigin( path ) ) {
-		if ( ! isPageRegistered() ) {
+		if ( ! isPageRegistered() && ! isPathOutOfScope( path ) ) {
```

Comment:
Maybe checking whether the current path is out of scope is enough. WDYT?

```js
if ( isCurrentPathOutOfScope( window.location.pathname ) ) {
 ...
} else {
  page.show();
}
```

Then, the following scenarios will work as expected
* `/setup/<path1>` to `/setup/<path2>` (using `window.history.pushState`)
* `/out-of-setup/<path>` to `/setup/<path>` (using `page.show`)

Reply:
> Maybe checking whether the current path is out of scope is enough. WDYT?

I think its enough to use ` ! isCurrentPathOutOfScope(path)`.
If the path is out of scope, it will use page.show, which is working for `/post` -> `/setup`
If the path is within scope, it will pushState, which is working for `/post/` -> `/home`

I don't have an example of `/setup/` -> `/setup` navigation since It seems like that navigation function is not specifically used in the stepper framework.

Reply:
Here is the sample case:
1. Go to `/setup?siteSlug=<your_free_site>`
2. Continue until you see `Pick a design` on the Design Picker screen
3. Select a premium theme
4. Unlock theme
5. Upgrade to activate
6. You have to see the checkout modal is open

Reply:
@arthur791004 - I pushed up some changes based on what we discussed. 
I tested out the following scenarios:

The design picker flow you described above.
`/setup/site-setup/designSetup` -> `/setup/site-setup/designSetup?siteSlug={SLUG}&products=premium` 

The post editor sidebar "Next steps" link to launchpad screen
`/post/SLUG/314` -> `/setup/.../launchpad`

The post editor sidebar "Dashboard" / "View posts" link
`/post/SLUG/314` -> `/home/SLUG`
`/post/SLUG/314` -> `/posts/SLUG`